sklearn/decomposition/_nmf.py

```diff
@@ -1177,6 +1113,10 @@ class NMF(TransformerMixin, BaseEstimator):
 
         .. versionadded:: 0.19
 
+    update_H : bool, default=True
+        Set to True, both W and H will be estimated from initial guesses.
+        Set to False, only W will be estimated.
+
```

Comment:
update_H is an argument of ``non_negative_factorization`` to distinguish between fit and transform: at transform you don't update H. We don't need it in the class. In the fit we do update H and in the tranform we don't.

Actually you need to update the code of ``transform`` because it still calls the function.

Also I think that we should deprecate ``update_H`` in the function since it was there for private purpose only. wdyt @TomDLT ?

Reply:
I agree we should not have `update_H` in the class.

To implement `update_H=False` in the function, we would need to either:
- create an NMF instance, fill the attributes normally added during the fit, and call transform
- or directly call the private solver functions
It makes the design a bit wonky, so I agree we could deprecate this parameter in the function.

Reply:
If the purpose of having the function version is to give users control of such thing, then removing it for the sake of a code cleanup might be throwing out the baby with the bath water. Is it not more appropriate to create a private class underlying NMF but also facilitating this need through the availability of extra parameters or controls?

Reply:
Fair point, it would remove a functionality, which is using the NMF `transform` without the `fit` first. I am not sure this is not something we generally want to support in scikit-learn, but if we want to keep this use-case, I think we should keep the current organization and not do the proposed refactoring.

Reply:
Perhaps the question is then not just whether update_H should be in the class, but whether it's useful to a subset of users, or whether it was only there for the convenience of the developer.

Reply:
My intuition is that this use case is not a non-negative matrix factorization use case but an ElasticNet problem with positivity constraint on the coefs which can actually be done (``ElasticNet(positive=True)``).

Reply:
If we want to keep this possibility we could add a ``H`` parameter to the transform method such that if it's not None, we allow to transform without having calling fit.

Reply:
The latest changes in the PR keep the behavior of update_H. The public API and functionalities aren't modified at all. I'll resolve the conversation now.